### PR TITLE
Clean-up sign-conversion warnings

### PR DESCRIPTION
**Pull request by 0x8000-0000**
I agree that my contributions are licensed under the {fmt} license, and agree to future changes to the licensing.

### CODE REVIEW

**Comment by 0x8000-0000** [comment]
This first step reduces the number of -Wsign-conversion warnings from 317 to 58 with GCC9.

**Comment by 0x8000-0000** [comment]
I have tried to reproduce the test crash on Visual Studio 2019 building in Release Mode for 32 bits (I don't have 2017)

    > "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\Common7\Tools\VsDevCmd.bat" -arch=x86

    > cmake ..\fmt -DFMT_PEDANTIC=ON -DCMAKE_BUILD_TYPE=Release -G"Visual Studio 16 2019"

    > cmake --build . --config Release

    > ctest -C Release

And I can't reproduce the crash. If somebody has access to that toolchain and could try debugging it, I would be very grateful. Otherwise I would have to do a binary search on all the changes in this commit :(

**Comment by 0x8000-0000** [comment]
Made doubly-sure I have produced 32 bit binaries with:

"C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake" ..\fmt -DFMT_PEDANTIC=ON -DCMAKE_BUILD_TYPE=Release -A win32

$ file bin/Release/color-test.exe
bin/Release/color-test.exe: PE32 executable (console) Intel 80386, for MS Windows


Still, no crash when running the 32-bit binary on Windows 64, running on AMD laptop (FWIW)

**Comment by 0x8000-0000** [comment]
This still crashes in UtilTest.WriteUIntPtr, even with just some changes in chrono.h ! I suspect the master branch just fails in the same way at this time.

**Comment by 0x8000-0000** [comment]
Oh, it "fails" in the same way as #1438 , even though that one just modifies a text document.

**Comment by vitaut** [comment]
Yeah, `WriteUIntPtr` test failure is unrelated. I'm about to fix it (just waiting for see if CI is green).

**Comment by vitaut** [comment]
Merged the fix into master. Please rebase.

**Comment by 0x8000-0000** [comment]
Down to 51 errors on GCC9, 18 on Clang9.

**Comment by 0x8000-0000** [comment]
Down to 1 warning in fmt code, reported by both Gcc9 and Clang9.

fmt/test/format-test.cc:2026:10: error: implicit conversion changes signedness: 'unsigned __int128' to '::testing::internal::Function<void (__int128_t)>::Argument1' (aka '__int128') [-Werror,-Wsign-conversion]

Gcc9 also reports several warnings stemming from a macro in test/gtest/gtest.h:

   1922 #define GTEST_COMPILE_ASSERT_(expr, msg) \                                                                                                             
   1923   typedef ::testing::internal::CompileAssert<(static_cast<bool>(expr))> \                                                                              
   1924       msg[static_cast<bool>(expr) ? 1 : -1] GTEST_ATTRIBUTE_UNUSED_            

I think there are some cases where Gcc9 sees "msg[-1]" and reports the warning. It might be worth exploring whether upgrading gtest/gmock would clean that up.

File: include/fmt/chrono.h [hunk 1]
```diff
@@ -793,11 +793,7 @@ struct chrono_formatter {
 
   explicit chrono_formatter(FormatContext& ctx, OutputIt o,
                             std::chrono::duration<Rep, Period> d)
-      : context(ctx), out(o), val(d.count()), negative(false) {
-    if (d.count() < 0) {
-      val = 0 - val;
-      negative = true;
-    }
+      : context(ctx), out(o), val(static_cast<rep>(std::abs(d.count()))), negative(d.count() < 0) {
```
Comment (vitaut) [inline include/fmt/chrono.h]: I suggest keeping the old code where negative case is handled in one place.

File: include/fmt/compile.h [hunk 2]
```diff
@@ -141,7 +141,7 @@ class format_string_compiler : public error_handler {
     auto it = parse_format_specs(begin, end, handler);
     if (*it != '}') on_error("missing '}' in format string");
     repl.arg_id = part_.part_kind == part::kind::arg_index
-                      ? arg_ref<Char>(part_.val.arg_index)
+                      ? arg_ref<Char>(static_cast<int>(part_.val.arg_index))
```
Comment (vitaut) [inline include/fmt/compile.h]: Please change the type of `arg_index` in `arg_ref` to `int` and remove the cast.

File: include/fmt/format-inl.h [hunk 3]
```diff
@@ -482,8 +483,8 @@ FMT_FUNC fp get_cached_power(int min_exponent, int& pow10_exponent) {
   const uint64_t one_over_log2_10 = 0x4d104d42;  // round(pow(2, 32) / log2(10))
   int index = static_cast<int>(
       static_cast<int64_t>(
-          (min_exponent + fp::significand_size - 1) * one_over_log2_10 +
-          ((uint64_t(1) << 32) - 1)  // ceil
+          static_cast<uint64_t>(min_exponent + fp::significand_size - 1) * one_over_log2_10 +
```
Comment (vitaut) [inline include/fmt/format-inl.h]: Please replace the cast with `to_unsigned`.

File: include/fmt/format-inl.h [hunk 4]
```diff
@@ -531,14 +532,14 @@ class bigint {
 
   friend struct formatter<bigint>;
 
-  void subtract_bigits(int index, bigit other, bigit& borrow) {
+  void subtract_bigits(size_t index, bigit other, bigit& borrow) {
```
Comment (vitaut) [inline include/fmt/format-inl.h]: In all the bigint and floating-point formatting code let's keep the type of indices `int`. More generally we don't need to address false positives in `format-inl.h` because it's an internal header.

File: include/fmt/format.h [hunk 5]
```diff
@@ -1244,7 +1244,7 @@ int snprintf_float(T value, int precision, float_specs specs,
                    buffer<char>& buf);
 
 template <typename T> T promote_float(T value) { return value; }
-inline double promote_float(float value) { return value; }
+inline double promote_float(float value) { return static_cast<double>(value); }
```
Comment (vitaut) [inline include/fmt/format.h]: Why is this needed?

File: include/fmt/format.h [hunk 6]
```diff
@@ -2627,11 +2627,11 @@ class format_string_checker {
   FMT_CONSTEXPR void on_text(const Char*, const Char*) {}
 
   FMT_CONSTEXPR void on_arg_id() {
-    arg_id_ = context_.next_arg_id();
+    arg_id_ = static_cast<unsigned>(context_.next_arg_id());
```
Comment (vitaut) [inline include/fmt/format.h]: This cast won't be needed with the change mentioned earlier.

File: include/fmt/format.h [hunk 7]
```diff
@@ -2627,11 +2627,11 @@ class format_string_checker {
...
     check_arg_id();
   }
   FMT_CONSTEXPR void on_arg_id(int id) {
-    arg_id_ = id;
+    arg_id_ = static_cast<unsigned>(id);
```
Comment (vitaut) [inline include/fmt/format.h]: ditto

File: include/fmt/printf.h [hunk 8]
```diff
@@ -399,10 +399,10 @@ template <typename OutputIt, typename Char>
 typename basic_printf_context<OutputIt, Char>::format_arg
 basic_printf_context<OutputIt, Char>::get_arg(unsigned arg_index) {
   if (arg_index == internal::max_value<unsigned>())
-    arg_index = parse_ctx_.next_arg_id();
+    arg_index = static_cast<unsigned>(parse_ctx_.next_arg_id());
```
Comment (vitaut) [inline include/fmt/printf.h]: Please change `arg_index` to `int` and remove all the casts.

File: test/format-impl-test.cc [hunk 9]
```diff
@@ -224,19 +224,23 @@ TEST(FPTest, Normalize) {
...
+      {static_cast<double>(1.0f), 0.9999999701976776, 1.0000000596046448},
       // min normal
-      {1.1754944e-38f, 1.1754942807573643e-38, 1.1754944208872107e-38},
+      {static_cast<double>(1.1754944e-38f), 1.1754942807573643e-38, 1.1754944208872107e-38},
       // max subnormal
-      {1.1754942e-38f, 1.1754941406275179e-38, 1.1754942807573643e-38},
+      {static_cast<double>(1.1754942e-38f), 1.1754941406275179e-38, 1.1754942807573643e-38},
       // min subnormal
-      {1e-45f, 7.006492321624085e-46, 2.1019476964872256e-45},
+      {static_cast<double>(1e-45f), 7.006492321624085e-46, 2.1019476964872256e-45},
```
Comment (vitaut) [inline test/format-impl-test.cc]: This is unrelated to sign conversion and adds unnecessary noise - please revert.

File: test/format-test.cc [hunk 10]
```diff
@@ -2005,9 +2005,11 @@ class mock_arg_formatter
     : public fmt::internal::arg_formatter_base<buffer_range> {
  private:
 #if FMT_USE_INT128
-  MOCK_METHOD1(call, void(__int128_t value));
+  MOCK_METHOD1(callSigned, void(__int128_t value));
+  MOCK_METHOD1(callUnsigned, void(__uint128_t value));
 #else
-  MOCK_METHOD1(call, void(long long value));
+  MOCK_METHOD1(callSigned, void(long long value));
+  MOCK_METHOD1(callUnsigned, void(unsigned long long value));
```
Comment (vitaut) [inline test/format-test.cc]: This is unnecessary for the test purposes. Just static cast to the largest signed type.

File: test/printf-test.cc [hunk 11]
```diff
@@ -589,7 +590,8 @@ class custom_printf_arg_formatter : public formatter_t {
       if (round(value * pow(10, specs()->precision)) == 0.0) value = 0;
   return formatter_t::operator()(value);
 }
-};
+}
+;
```
Comment (vitaut) [inline test/printf-test.cc]: Could you remove ; while at it? =)

File: test/std-format-test.cc [hunk 12]
```diff
@@ -110,7 +110,7 @@ template <> struct std::formatter<S> {
     char c = get_char();
     if (!isdigit(c) || (++iter, get_char()) != '}')
       throw format_error("invalid format");
-    width_arg_id = c - '0';
+    width_arg_id = static_cast<size_t>(c - '0');
```
Comment (vitaut) [inline test/std-format-test.cc]: This tests an example from the standard and should remain unchanged.

File: include/fmt/chrono.h (hunk 1 above)
Comment (0x8000-0000) [reply include/fmt/chrono.h]: You are trying to initialize 'val' which is unsigned with a signed value (d.count()).

File: include/fmt/compile.h (hunk 2 above)
Comment (0x8000-0000) [reply include/fmt/compile.h]: It seems to be already 'int' in struct format_part.value . Will remove the cast.

File: include/fmt/format-inl.h (hunk 3 above)
Comment (0x8000-0000) [reply include/fmt/format-inl.h]: Your suggested change causes four tests failures (building with Gcc9 in Debug mode):

          5 - grisu-test (Child aborted)
          7 - format-test (Child aborted)
          8 - format-impl-test (Failed)
         12 - printf-test (Child aborted)

It trips this:

    fmt/include/fmt/core.h:266: assertion failed: negative value

File: include/fmt/format-inl.h (hunk 4 above)
Comment (0x8000-0000) [reply include/fmt/format-inl.h]: The code is trying to subscript into bigits object (of type buffer) with a signed value, when the operator [] expects std::size_t, thus the warning.

Do you want me to convert the buffer implementation and interface from size_t to int? This, after all, what Scott Meyers and Bjarne Stroustroup are recommending:
     http://open-std.org/JTC1/SC22/WG21/docs/papers/2019/p1428r0.pdf
     https://www.aristeia.com/Papers/C++ReportColumns/sep95.pdf

Converting to int from size_t *should* clean up a lot of casts.

File: include/fmt/format.h (hunk 5 above)
Comment (0x8000-0000) [reply include/fmt/format.h]: Seems it is not needed. Removed.

File: include/fmt/printf.h (hunk 8 above)
Comment (0x8000-0000) [reply include/fmt/printf.h]: This will be a bit tricky because of line 401. Should I change the sentinel value to internal::max_value<int>() ?

File: test/std-format-test.cc (hunk 12 above)
Comment (0x8000-0000) [reply test/std-format-test.cc]: Should width_arg_id become signed then?

File: test/format-test.cc (hunk 10 above)
Comment (0x8000-0000) [reply test/format-test.cc]: The test is calling the mocked method with both signed and unsigned values.

I don't care as much about the tests being -Wsign-conversion clean, but...

I want my own code to be -Wsign-conversion clean, because we got bitten in the past by (unsigned)-1, and other bit fields confusion (somebody shoving 32 bits into a 16 bit register) and I want to use {fmt}. Which means {fmt} needs to be -Wsign-conversion clean, which means its test suite needs to be -Wsign-conversion clean, otherwise we can't enforce it via the automated regression suite.

File: test/format-impl-test.cc (hunk 9 above)
Comment (0x8000-0000) [reply test/format-impl-test.cc]: OK

File: test/printf-test.cc (hunk 11 above)
Comment (0x8000-0000) [reply test/printf-test.cc]: Can't, since it follows a class definition :)  Not sure why clang-format wants to put it on a separate line.

File: include/fmt/chrono.h (hunk 1 above)
Comment (vitaut) [reply include/fmt/chrono.h]: Adding a cast is fine. I'm just saying that having negation and setting `negative` in one if block makes the intent a bit clearer, so let's keep that part.

File: include/fmt/format-inl.h (hunk 3 above)
Comment (vitaut) [reply include/fmt/format-inl.h]: This suggests that the cast is misplaced.

File: include/fmt/format.h (hunk 6 above)
Comment (0x8000-0000) [reply include/fmt/format.h]: This is a different arg_id_ ; do you want me to convert it to 'int' as well?

File: include/fmt/chrono.h (hunk 1 above)
Comment (0x8000-0000) [reply include/fmt/chrono.h]: OK

File: include/fmt/format-inl.h (hunk 4 above)
Comment (vitaut) [reply include/fmt/format-inl.h]: Changing `buffer`'s index to a signed type might be cleaner but let's not do this in the current PR. For now let's just keep floating-point related code as is.

File: include/fmt/printf.h (hunk 8 above)
Comment (vitaut) [reply include/fmt/printf.h]: I think sentinel can be changed to -1 once `arg_index` is signed.

File: include/fmt/format-inl.h (hunk 3 above)
Comment (0x8000-0000) [reply include/fmt/format-inl.h]: Yes, in fact it is not needed at all.

Comment (0x8000-0000) [reply include/fmt/format-inl.h]: Removing the cast also silences the warning, but still trips the assertion.

File: include/fmt/format-inl.h (hunk 4 above)
Comment (0x8000-0000) [reply include/fmt/format-inl.h]: OK, will leave changing buffer for a follow-up change.

File: include/fmt/printf.h (hunk 8 above)
Comment (0x8000-0000) [reply include/fmt/printf.h]: I was thinking of wrapping the sentinel value in a constant, introduced via an enum so it is portable to older compilers.

File: test/format-test.cc (hunk 10 above)
Comment (vitaut) [reply test/format-test.cc]: This is a simple mock which is only used here: https://github.com/fmtlib/fmt/blob/62da1db62affe82dc3ff5e5396e7b5cc235eca3d/test/format-test.cc#L2052

so casting is fine.

Comment (0x8000-0000) [reply test/format-test.cc]: I changed as suggested and I got back the dreaded:

fmt/test/format-test.cc:2026:10: error: implicit conversion changes signedness: 'unsigned __int128' to '::testing::internal::Function<void (__int128_t)>::Argument1' (aka '__int128') [-Werror,-Wsign-conversion]

File: test/std-format-test.cc (hunk 12 above)
Comment (0x8000-0000) [reply test/std-format-test.cc]: width_arg_id is passed on the next line to check_arg_id which requires a size_t .

File: test/format-test.cc (hunk 10 above)
Comment (0x8000-0000) [reply test/format-test.cc]: This is the long story, threaded through format.h and core.h:

> /home/florin/libraries/fmt/test/format-test.cc: In instantiation of ‘typename std::enable_if<fmt::v6::internal::is_integral<T>::value, std::back_insert_iterator<fmt::v6::internal::buffer<char> > >::type mock_arg_formatter::operator()(T) [with T = __int128 unsigned; typename std::enable_if<fmt::v6::internal::is_integral<T>::value, std::back_insert_iterator<fmt::v6::internal::buffer<char> > >::type = std::back_insert_iterator<fmt::v6::internal::buffer<char> >]’:

> /home/florin/libraries/fmt/include/fmt/core.h:1024:15:   required from ‘constexpr decltype (vis(0)) fmt::v6::visit_format_arg(Visitor&&, const fmt::v6::basic_format_arg<Context>&) [with Visitor = mock_arg_formatter; Context = fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char>; decltype (vis(0)) = std::back_insert_iterator<fmt::v6::internal::buffer<char> >]’

> /home/florin/libraries/fmt/include/fmt/format.h:3172:25:   required from ‘void fmt::v6::format_handler<ArgFormatter, Char, Context>::on_replacement_field(const Char*) [with ArgFormatter = mock_arg_formatter; Char = char; Context = fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char>]’

> /home/florin/libraries/fmt/include/fmt/format.h:2583:7:   required from ‘constexpr void fmt::v6::internal::parse_format_string(fmt::v6::basic_string_view<Char>, Handler&&) [with bool IS_CONSTEXPR = false; Char = char; Handler = fmt::v6::format_handler<mock_arg_formatter, char, fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char> >&]’

> /home/florin/libraries/fmt/include/fmt/format.h:3205:39:   required from ‘typename Context::iterator fmt::v6::vformat_to(typename ArgFormatter::range, fmt::v6::basic_string_view<Char>, fmt::v6::basic_format_args<Context>, fmt::v6::internal::locale_ref) [with ArgFormatter = mock_arg_formatter; Char = char; Context = fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char>; typename Context::iterator = std::back_insert_iterator<fmt::v6::internal::buffer<char> >; typename ArgFormatter::range = fmt::v6::buffer_range<char>]’

> /home/florin/libraries/fmt/test/format-test.cc:2043:63:   required from here
> /home/florin/libraries/fmt/test/format-test.cc:2026:10: error: conversion to ‘testing::internal::Function<void(__int128)>::Argument1’ {aka ‘__int128’} from ‘__int128 unsigned’ may change the sign of the result [-Werror=sign-conversion]
 2026 |     call(value);
      |          ^~~~~

**Comment by 0x8000-0000** [comment]
Once this is approved I can refactor the top commit and push the changes into the earlier commits. I didn't want to force-push during the review.

File: include/fmt/format-inl.h (hunk 3 above)
Comment (0x8000-0000) [reply include/fmt/format-inl.h]: Actually the static_cast<int64_t> is not needed.

File: include/fmt/format.h (hunk 6 above)
Comment (0x8000-0000) [reply include/fmt/format.h]: Done.

**Comment by vitaut** [comment]
I guess this can be closed now in favor of #1440.
